xds/src/main/java/com/linecorp/centraldogma/xds/internal/CentralDogmaSnapshotResources.java

```diff
@@ -0,0 +1,109 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.centraldogma.xds.internal;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.hash.Hashing;
+import com.google.protobuf.Message;
+
+import com.linecorp.centraldogma.common.Revision;
+
+import io.envoyproxy.controlplane.cache.ResourceVersionResolver;
+import io.envoyproxy.controlplane.cache.Resources;
+import io.envoyproxy.controlplane.cache.SnapshotResources;
+import io.envoyproxy.controlplane.cache.VersionedResource;
+
+final class CentralDogmaSnapshotResources<T extends Message> extends SnapshotResources<T> {
+
+    public static <T extends Message> SnapshotResources<T> create(
+            Iterable<T> resources, Revision revision) {
+        final ImmutableMap.Builder<String, VersionedResource<T>> versionedResourcesMap = ImmutableMap.builder();
+        final ImmutableMap.Builder<String, T> resourcesMap = ImmutableMap.builder();
+        for (T resource : resources) {
+            final String resourceName = Resources.getResourceName(resource);
+            versionedResourcesMap.put(resourceName, VersionedResource.create(resource));
+            resourcesMap.put(resourceName, resource);
+        }
+        return new CentralDogmaSnapshotResources<>(versionedResourcesMap.build(), resourcesMap.build(),
+                                                   Integer.toString(revision.major()));
+    }
+
+    private final Map<String, VersionedResource<T>> versionedResources;
+    private final Map<String, T> resources;
+    private final ResourceVersionResolver resourceVersionResolver;
+
+    private CentralDogmaSnapshotResources(
+            Map<String, VersionedResource<T>> versionedResources, ImmutableMap<String, T> resources,
+            String allResourceVersion) {
+        this.versionedResources = versionedResources;
+        this.resources = resources;
+        resourceVersionResolver = resourceNames -> {
+            if (resourceNames.isEmpty()) { // All resources
+                return allResourceVersion;
+            }
+            if (resourceNames.size() == 1) {
+                final VersionedResource<T> versionedResource = versionedResources.get(resourceNames.get(0));
+                if (versionedResource == null) {
+                    return "";
+                }
+                return versionedResource.version();
+            }
+            final ArrayList<String> sorted = new ArrayList<>(resourceNames);
+            sorted.sort(String::compareTo);
+
+            final List<VersionedResource<T>> collected = sorted.stream().map(versionedResources::get)
```

Comment:
Note: it is fine to omit resources that are missing when calculating the version because the server will also take account the missing resources when determining how to create a response